controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -572,7 +572,10 @@ private boolean verifyRecordMatchesInput(List<Long> segmentsToSeal, List<SimpleE
 
                                         final SegmentRecord latestSegment = TableHelper.getLatestSegmentRecord(segmentIndex.getData(),
                                                 segmentTable.getData());
-                                        if (latestSegment.getCreationEpoch() < newEpoch) {
+                                        final HistoryRecord activeEpoch = TableHelper.getActiveEpoch(historyIndex.getData(),
+                                                historyTable.getData());
+
+                                        if (latestSegment.getCreationEpoch() < newEpoch && activeEpoch.getEpoch() == epochTransition.getActiveEpoch()) {
```

Comment:
I don't understand this. Why would the history's active epoch be different from the epoch transition's active epoch? If this is possible due to concurrency, then I don't understand how it can't happen right after I perform the if check, which would be equally problematic.

Reply:
Rolling transaction :)

Scale and rolling txn now run on different event-processors so they both independently create processing specific records (EpochTransitionRecord or CommittingTxnRecord respectively) followed by attempting to start processing by setting the state to either SCALING or COMMITTING_TXN. Whichever is able to set the state gets to start its processing while the other is rescheduled. 

So if after creating epoch transition record, a rollingtransaction happens, then when scale's processing starts, its epoch-transition may have become stale and it should be rejected here and the record cleaned up (that is done in isEpochTransitionConsistent method).